Nodejs/Product/Nodejs/SharedProject/CommonProjectNode.DiskMerger.cs

```diff
@@ -12,128 +15,226 @@ internal abstract partial class CommonProjectNode
+            private readonly BlockingCollection<DirState> dirs = new BlockingCollection<DirState>(new ConcurrentStack<DirState>());
+
             private readonly CommonProjectNode project;
 
-            public DiskMerger(CommonProjectNode project, HierarchyNode parent, string dir)
+            private readonly Thread worker;
+
+            private readonly ManualResetEventSlim processingDirsEvent = new ManualResetEventSlim(initialState: true);
+
+            private bool isDisposed = false;
+
+            public DiskMerger(CommonProjectNode project)
             {
                 this.project = project;
-                this.remainingDirs.Push(new DirState(dir, parent));
+
+                this.worker = new Thread(this.Merge)
+                {
+                    Name = "NodeTools Project Disk Sync",
+                    IsBackground = true,
+                };
             }
 
-            /// <summary>
-            /// Continues processing the merge request, performing a portion of the full merge possibly
-            /// returning before the merge has completed.
-            /// 
-            /// Returns true if the merge needs to continue, or false if the merge has completed.
-            /// </summary>
-            public bool ContinueMerge(bool hierarchyCreated)
+            public void Start()
             {
-                if (this.remainingDirs.Count == 0)
+                if (this.isDisposed || this.dirs.IsAddingCompleted)
                 {
-                    // all done
-                    this.project.BoldStartupItem();
-                    return false;
+                    throw new ObjectDisposedException(nameof(DiskMerger));
                 }
 
-                var dir = this.remainingDirs.Pop();
-                if (!Directory.Exists(dir.Name))
+                if (!this.worker.IsAlive)
                 {
-                    return true;
+                    this.worker.Start();
                 }
+            }
 
-                var wasExpanded = hierarchyCreated ? dir.Parent.GetIsExpanded() : false;
-                var missingChildren = new HashSet<HierarchyNode>(dir.Parent.AllChildren);
-                try
+            public void AddDir(HierarchyNode parent, string folderPath)
+            {
+                if (this.isDisposed || this.dirs.IsAddingCompleted)
                 {
-                    foreach (var curDir in Directory.EnumerateDirectories(dir.Name))
+                    throw new ObjectDisposedException(nameof(DiskMerger));
+                }
+                this.dirs.Add(new DirState(folderPath, parent));
+            }
+
+            public bool CompletedMerge => this.dirs.Count == 0;
+
+            public void Wait()
+            {
+                if (!this.isDisposed)
+                {
+                    this.processingDirsEvent.Wait();
+                }
+            }
+
+            private void Merge()
+            {
+                while (!this.isDisposed && !this.dirs.IsCompleted)
```

Comment:
Here, and in other places (e.g. "AddDir" above), you have no lock around checking `isDisposed`, and then accessing `this.dirs`. Thus is could be disposed in-between the check and the access in multi-threaded scenarios (which this is), causing occasional race-conditions/crashes.